src/parser/ui/ActiveTimeGraph.tsx

```diff
@@ -2,24 +2,111 @@ import BaseChart, { formatTime } from 'parser/ui/BaseChart';
 import { AutoSizer } from 'react-virtualized';
 import { VisualizationSpec } from 'react-vega';
 
-export interface UptimeHistoryEntry {
-  timestamp: number;
-  uptimePct: number;
-}
+/** The maximum value on the Y axis. Can't have active time higher than 100%. */
+const GRAPH_MAX_Y = 1;
+/** Minimum height of graph */
+const GRAPH_HEIGHT = 160;
+/** Sample period, in milliseconds */
+const SAMPLE_PERIOD = 3000;
+/** Duration of rolling average window, in milliseconds */
+const ROLLING_AVERAGE_WINDOW_DURATION = 9000;
 
+/** Data point effectively defines x (time) and y (active time percent) axes */
 interface GraphData {
   timestamp: number;
   activeTimePercentage: number;
 }
 
-const GRAPH_SAMPLE_COUNT = 2; // Show value for every Nth GCD in the graph.
-const GRAPH_MAX_Y = 1; // The maximum value on the Y axis. Used to prevent the graph skyrocketing.
+type Props = {
+  activeTimeSegments: { start: number; end: number }[];
+  fightStart: number;
+  fightEnd: number;
+};
 
-export default function getUptimeGraph(
-  uptimeHistory: UptimeHistoryEntry[],
-  fightStartTime: number,
+const ActiveTimeGraph = ({ activeTimeSegments, fightStart, fightEnd, ...others }: Props) => {
+  // Generate active time rolling average from the ABC module's active time segments
+  const graphData: GraphData[] = [];
+  let currIdx = 0;
+  for (
+    let timestamp = fightStart + SAMPLE_PERIOD;
+    timestamp < fightEnd;
+    timestamp += SAMPLE_PERIOD
+  ) {
+    // advance segment index until we've passed window
+    while (currIdx < activeTimeSegments.length && activeTimeSegments[currIdx].start < timestamp) {
+      currIdx += 1;
+    }
+    const windowDuration = Math.min(ROLLING_AVERAGE_WINDOW_DURATION, timestamp - fightStart);
+    const activeTimePercentage = getRollingAverage(
+      activeTimeSegments,
+      currIdx - 1,
+      timestamp,
+      windowDuration,
+    );
+    graphData.push({ timestamp, activeTimePercentage });
+  }
```

Comment:
you should use `useMemo` to keep re-renders from re-running this code. (not a correctness issue, just uses less CPU). something like this:

```tsx
  // Generate active time rolling average from the ABC module's active time segments
  const graphData = useMemo(() => {
    const graphData: GraphData[] = [];
    // existing code (github suggestion diff is terrible)
    return graphData;
  }, [fightStart, fightEnd, activeTimeSegments]);
```

Reply:
Yeah good idea I'll add that